sysbuild: Add support for COMPILER_WARNINGS_AS_ERRORS

nordicjm opened the pull request:
Adds support for the Kconfig option to enable compiler warnings being treated as errors in all images that are build. This is a sticky-set option which means enabling it in sysbuild will enable it in all images, if it is then disabled in sysbuild it will not be disabled in any of the images as a result and would need to be manually unselected.

Fixes #67360
Fixes #70273

nordicjm committed:
sysbuild: Add support for COMPILER_WARNINGS_AS_ERRORS

Adds support for the Kconfig option to enable compiler warnings
being treated as errors in all images that are build. This is a
sticky-set option which means enabling it in sysbuild will
enable it in all images, if it is then disabled in sysbuild it
will not be disabled in any of the images as a result and would
need to be manually unselected.

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

tejlmand commented on share/sysbuild/image_configurations/BOOTLOADER_image_default.cmake:
```diff
@@ -27,3 +27,5 @@ foreach(loopkeytype ${keytypes})
   endif()
 endforeach()
+
+include(image_configurations/ALL_image_default.cmake)
```
No need to place this line here.

The `IMAGE_CONF_SCRIPT` property supports lists exactly for the purpose of allowing multiple conf scripts.
https://github.com/zephyrproject-rtos/zephyr/blob/f34f06058890e8a2628429bf67917dc8d6996487/share/sysbuild/cmake/modules/sysbuild_extensions.cmake#L456-L458
https://github.com/zephyrproject-rtos/zephyr/blob/f34f06058890e8a2628429bf67917dc8d6996487/share/sysbuild/cmake/modules/sysbuild_extensions.cmake#L472

So just add `ALL_image_default.cmake` to that property for all images, regardless of boot / main images.

```suggestion
```

nordicjm replied:
Fixed

tejlmand commented on share/sysbuild/cmake/modules/sysbuild_extensions.cmake:
```diff
@@ -375,6 +375,9 @@ function(ExternalZephyrProject_Add)
   if(DEFINED ZBUILD_APP_TYPE)
     set(image_default "${CMAKE_SOURCE_DIR}/image_configurations/${ZBUILD_APP_TYPE}_image_default.cmake")
     set_target_properties(${ZBUILD_APPLICATION} PROPERTIES IMAGE_CONF_SCRIPT ${image_default})
+  else()
+    set(image_default "${CMAKE_SOURCE_DIR}/image_configurations/ALL_image_default.cmake")
+    set_target_properties(${ZBUILD_APPLICATION} PROPERTIES IMAGE_CONF_SCRIPT ${image_default})
   endif()
```
Just append for all images:
```suggestion
  endif()
  set(image_default "${CMAKE_SOURCE_DIR}/image_configurations/ALL_image_default.cmake")
  set_property(TARGET ${ZBUILD_APPLICATION} APPEND PROPERTY IMAGE_CONF_SCRIPT ${image_default})
```

tejlmand commented on share/sysbuild/image_configurations/MAIN_image_default.cmake:
```diff
@@ -20,3 +20,5 @@ if(SB_CONFIG_BOOTLOADER_MCUBOOT)
   endif()
 endif()
+
+include(image_configurations/ALL_image_default.cmake)
```
see other comment
```suggestion
```

nordicjm replied:
Fixed

tejlmand commented on share/sysbuild/Kconfig:
```diff
@@ -41,4 +41,9 @@ config WARN_DEPRECATED
 	  Print a warning when the Kconfig tree is parsed if any deprecated
 	  features are enabled.
 
+config COMPILER_WARNINGS_AS_ERRORS
+	bool "Treat warnings as errors"
+	help
+	  Turn on "warning as error" toolchain flags for all images if set.
+
```
Proposal, not required.

Should we already now place this in a build (tools) / compiler specific Kconfig file ?
I suspect this won't be the last setting of this kind.

57300 replied:
My proposal would be to consider #66253 and using `-Dall_CONFIG_COMPILER_WARNINGS_AS_ERRORS=y`.
In general, I don't know if it makes sense for sysbuild to control toolchain settings.

tejlmand replied:
> My proposal would be to consider https://github.com/zephyrproject-rtos/zephyr/pull/66253 and using -Dall_CONFIG_COMPILER_WARNINGS_AS_ERRORS=y.

yes, that would be a good solution for this specific use-case.
But we still have some outstanding issues related to #66253, my biggest concern related to Kconfig settings which takes a relative path. Relative in context of an `all_CONFIG_SETTING` would be problematic.

So until we have decided how to solve such cases, then I believe this PR is an acceptable solution to the issue at hand.

> In general, I don't know if it makes sense for sysbuild to control / be aware of toolchain settings.

I agree there.
Especially because different images could be building with different compilers / architecture specific flags, and therefore would be target (image) specific settings.

But there are few settings, like warnings as errors, which are generic enough and thus make sense to share.

nordicjm replied:
Moved to `build` folder

PerMac committed:
twister: Use proper warnings-to-errors command for sysbuild

If sysbuild is used a flag SB_CONFIG_COMPILER_WARNINGS_AS_ERRORS
has to be used in order to turn warings to errors on all images.
Align pytest tests with the change.

Fixes: #67360

Signed-off-by: Maciej Perkowski <Maciej.Perkowski@nordicsemi.no>

nordicjm commented:
@nordic-krch there is a problem with `samples/subsys/logging/multidomain` see build log and submit a fix please

nordicjm committed:
samples: ipc: rpmsg_service: Add sample file, remove broken builds

Adds a sample file so that this sample is built in CI and removes
all configurations that do not build

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

tejlmand commented on samples/subsys/ipc/rpmsg_service/Kconfig.sysbuild:
```diff
@@ -6,11 +6,5 @@ source "share/sysbuild/Kconfig"
 	string
-	default "nrf5340dk/nrf5340/cpunet" if $(BOARD) = "nrf5340dk"
-	default "bl5340_dvk/nrf5340/cpunet" if $(BOARD) = "bl5340_dvk"
-	default "lpcxpresso54114/lpc54114/m0" if $(BOARD) = "lpcxpresso54114"
-	default "mps2_an521_remote" if $(BOARD) = "mps2_an521"
-	default "v2m_musca_b1_ns" if $(BOARD) = "v2m_musca_b1"
-	default "esp32_devkitc_wroom_appcpu" if $(BOARD) = "esp32_devkitc_wroom"
-	default "esp32_devkitc_wrover_appcpu" if $(BOARD) = "esp32_devkitc_wrover"
-	default "esp32s3_devkitm_appcpu" if $(BOARD) = "esp32s3_devkitm"
```
not good to have boards present which doesn't work for this sample, but afaik this used to work before the sysbuild transition here: https://github.com/zephyrproject-rtos/zephyr/pull/65072

@sylvioalves can you take a look at why this doesn't work anymore ?

sylvioalves replied:
Checking.

sylvioalves replied:
@nordicjm @tejlmand LGTM. Better to remove this boards while I finish performing necessary changes to use sysbuild here. This might take a few days due to other stuff. Solution used prior HWMv2 does not fit here anymore.

tejlmand reviewed (commented):
Not happy about removing files which might be needed later if the removed boards need to be supported again.

Let's give @sylvioalves a chance to fix the boards or provide a time-frame when supported can be re-added before going forward.
